### DISCUSSION TITLE
Implement multiple Keys search using views

### DISCUSSION THREAD

**Discussion by subin94** [apache/couchdb#3029] [Q&A] [answered]
Map Function

`function (doc) {
  emit([doc.key1, doc.key2, doc.key3, doc.key4], 1)
}
`

when I tried `?startkey=["key1-value","key2-value","key3-value"]&endkey=["key1-value","key2-value","key3-value"]` got the expected result, as filtered by key1, key2, key3

when tried with `?startkey=["key1-value","a","key3-value"]&endkey=["key1-value",{},"key3-value"]` was expecting the result to be filtered by key1 and key3 alone but filter based on key3 didnt happen.

In the given map function, if I need to perform a search using key3 or key4 alone, is it possible ?

Any help here is much appreciated 😅

**Comment by Terreii** [comment, marked as answer]
Map views can't be used to search in this way. You would have to create a second view, where key2 and key3 are swapped:

```javascript
function (doc) { emit([doc.key1, doc.key3, doc.key2, doc.key4], 1) }
```

#### Longer answer:
Map/Reduce views are for fast lookup and retrieving of data. To achieve that CouchDB stores your emitted values in a B-Tree, which acts like a list, sorted by their keys.

key1 values | key2 values | key3 values | key4 values
------------|-------------|-------------|-------------
a | a | a | a
a | a | a | b
a | a | b | a
a | a | b | b
a | b | a | a

When you now query with a startkey and endkey, will CouchDB look in that list, where each key would be. And then it will return to you all rows from startkey to endkey and all between them.

With a startkey of ["a", "a", "b"] and endkey of ["a", {}, "b"]

key1 values | key2 values | key3 values | key4 values | Key would be | Returned
------------|-------------|-------------|-------------|------|----
a | a | a | a | | no
a | a | a | b | | no
a | a | b | nothing (before null) | ⬅︎ startkey would be here | yes, if it would exist
a | a | b | a | | yes
a | a | b | b | | yes
a | b | a | a | | yes
a | {} | b | ... | ⬅︎ endkey would be here | yes, if it would exist and `inclusive_end` is not `false`

CouchDB returns slices of the list/B-tree.

**Comment by rjharmon** [comment]
I will add a useful analogy: your request would be like asking me to find words in the dictionary whose second letter is an E.  Because the typical Webster's (or any other brand of) dictionary isn't ordered according to second-letters, its data structure can't help me with the task.  Same way with the index in your example, @subin94
